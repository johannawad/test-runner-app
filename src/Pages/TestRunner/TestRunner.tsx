import React, { Dispatch } from "react";
import { TestItem } from "../../Components/TestItem/TestItem";
import { Test, TestStatus } from "../../Types/Test";
import { connect } from "react-redux";
import { Button, Container } from "../../Styles/general.styles";
import {
  fetchTestsRequestAction,
  runAllAction,
} from "../../Actions/tests.actions";

interface TestRunnerComponentProps {
  fetchTestsRequest: () => void;
  runAllTests: () => void;
  loading?: boolean;
  tests?: Test[];
}
const TestRunner = ({
  tests,
  loading,
  runAllTests,
  fetchTestsRequest,
}: TestRunnerComponentProps) => {
  React.useEffect(() => {
    fetchTestsRequest();
  }, [fetchTestsRequest]);

  const areTestsCompleted: boolean = tests
    ? tests.every(
        (t) => t.status === TestStatus.Failed || t.status === TestStatus.Passed
      )
    : false;

  return (
    <Container>
      {loading && <p>...loading</p>}
      <h1>Tests Runner 🏃🏽‍♀️</h1>
      {tests &&
        tests.map((test: Test) => (
          <TestItem key={tests.indexOf(test)} test={test} />
        ))}
      {areTestsCompleted && <h2>Tests Completed!</h2>}
      <Button data-testid="btn-run-tests" onClick={runAllTests}>
        Run tests
      </Button>
    </Container>
  );
};

const mapStateToProps = (state: any) => {
  return {
    tests: state.tests.data,
    loading: state.tests.loading,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  fetchTestsRequest: (): void => dispatch(fetchTestsRequestAction()),
  runAllTests: (): void => dispatch(runAllAction()),
});
export default connect(mapStateToProps, mapDispatchToProps)(TestRunner);
