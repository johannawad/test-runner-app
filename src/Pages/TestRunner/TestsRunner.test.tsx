import { shallow } from "enzyme";
import React from "react";
import TestRunner from "./TestRunner";
import configureStore from "redux-mock-store";
import { fireEvent, render } from "@testing-library/react";

const mockStore = configureStore([]);
describe("TestRunner", () => {
  let store: any;
  let initialState: any;

  beforeEach(() => {
    initialState = { tests: { data: [], loading: false, errors: undefined } };
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  it("should match snapshot", () => {
    const component = shallow(<TestRunner store={store} />);
    expect(component).toMatchSnapshot();
  });

  it("should dipatch @@tests/FETCH_REQUEST on mount", () => {
    render(<TestRunner store={store} />);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: "@@tests/FETCH_REQUEST",
    });
  });

  it("should dipatch @@tests/RUN_ALL on button click", () => {
    const component = render(<TestRunner store={store} />);
    const btn = component.getByTestId("btn-run-tests");

    fireEvent.click(btn);

    expect(store.dispatch).toHaveBeenCalledWith({
      type: "@@tests/RUN_ALL",
    });
  });
});
