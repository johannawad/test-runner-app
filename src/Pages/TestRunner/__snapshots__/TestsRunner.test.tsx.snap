// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestRunner should match snapshot 1`] = `
<ContextProvider
  value={null}
>
  <TestRunner
    fetchTestsRequest={[Function]}
    loading={false}
    runAllTests={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    tests={Array []}
  />
</ContextProvider>
`;
